helm: split proxy/auth documentation (#19881)
gravitational/teleport: docs/pages/reference/helm-reference/teleport-cluster.mdx

description: Values that can be set using the teleport-cluster Helm chart
---

The `teleport-cluster` Helm chart deploys a Teleport cluster on Kubernetes.

> This includes deploying proxies, auth servers, kubernetes-access.

Reviewer: Un-resolving for the question at the end.

> This includes deploying proxies, auth servers, kubernetes-access.

Author: I wanted to say "A Teleport instance running the kube_service and providing Kubernetes Access to the cluster". But it felt weird to list both Teleport component (proxies and auth servers) and Teleport features on the same level.

I'll switch to Kubernetes Access with the link

See the [Teleport HA Architecture page](../../deploy-a-cluster/high-availability.mdx) for more details.

The chart source is available
[on GitHub](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-cluster).

> The `teleport-cluster` chart runs three Teleport services split in two sets of pods:

Reviewer 2: ```suggestion
The `teleport-cluster` chart runs three Teleport services, split into two sets of pods:
```

<Admonition type="tip" title="Additional Kubernetes Clusters and Teleport Services">
If you want to provide access to resources like Databases, Applications or other
Kubernetes clusters than the one hosting the Teleport cluster, you should use the
[`teleport-kube-agent` Helm chart](teleport-kube-agent.mdx).

- `teleport-cluster` hosts a Teleport cluster, you should only need one.
- `teleport-kube-agent` connects to an existing Teleport cluster and exposes configured resources.
</Admonition>

The `teleport-cluster` chart can be deployed in four different modes. Get started with a guide for each mode:

| `chartMode` | Purpose | Guide |
| - | - | - |
| `standalone` | Runs by relying only on Kubernetes resources. | [Getting Started - Kubernetes with SSO](../../deploy-a-cluster/helm-deployments/kubernetes-cluster.mdx) |
| `aws` | Leverages AWS managed services to store data. | [Running an HA Teleport cluster using an AWS EKS Cluster](../../deploy-a-cluster/helm-deployments/aws.mdx) |
| `gcp` | Leverages GCP managed services to store data. | [Running an HA Teleport cluster using a Google Cloud GKE cluster](../../deploy-a-cluster/helm-deployments/gcp.mdx) |

> | `scratch` (v12 and above) | Generates empty Teleport configuration. User must pass its own config.  | [Running a Teleport cluster with a custom config](../../deploy-a-cluster/helm-deployments/custom.mdx) |

Reviewer 2: ```suggestion
| `scratch` (v12 and above) | Generates empty Teleport configuration. User must pass their own config. | [Running a Teleport cluster with a custom config](../../deploy-a-cluster/helm-deployments/custom.mdx) |
```

<Admonition type="note" title="Custom mode removal">

> `custom` mode has been removed with Teleport version 12. See the [version 12 migration guide]().

Reviewer 2: ```suggestion
`custom` mode has been removed in Teleport version 12. See the [version 12 migration guide]().
```

Also fix link

</Admonition>

<Admonition type="warning" title="Version Compatibility">
The chart is versioned with Teleport. No compatibility guarantees are ensured
between new charts and previous major Teleport versions. It is strongly recommended
to always deploy a Teleport version with the same major version as the Helm chart.
</Admonition>

> This reference details available values for the `teleport-cluster` chart.

Author: I moved the sentence up

(!docs/pages/includes/backup-warning.mdx!)

## `clusterName`

| Type | Default value | Required? | `teleport.yaml` equivalent |
| - | - | - | - |
| `string` | `nil` | Yes | `auth_service.cluster_name`, `proxy_service.public_addr` |

`clusterName` controls the name used to refer to the Teleport cluster, along with
the externally-facing public address to use to access it. In most setups this must
be a fully-qualified domain name (e.g. `teleport.example.com`) as this values is
used as the cluster's public address by default.

<Admonition type="note">
  When using a fully qualified domain name as your `clusterName`, you will also
  need to configure the DNS provider for this domain to point to the external
  load balancer address of your Teleport cluster.

  (!docs/pages/kubernetes-access/helm/includes/kubernetes-externaladdress.mdx!)

  You will need to manually add a DNS A record pointing `teleport.example.com` to
  either the IP or hostname of the Kubernetes load balancer.

  <Details title="Using Application Access?">
    (!docs/pages/includes/dns-app-access.mdx!)
  </Details>
</Admonition>

<Admonition type="warning">

>   The clusterName must not change in the Teleport cluster lifespan.

Reviewer 2: ```suggestion
  The `clusterName` cannot be changed during a Teleport cluster's lifespan. If you need to change it, you must redeploy a completely new cluster.
```

| - | - | - | - |
| `string` | `clusterName` value | no | `kubernetes_service.kube_cluster_name` |

`kubeClusterName` sets the name used for Kubernetes access.
This name will be shown to Teleport users connecting to the Kubernetes cluster.

## `auth`
| Type     | Default value | Required? |
|----------|---------------|-----------|
| `object` | {}            | no        |

The `teleport-cluster` chart deploys two sets of pods: auth and proxy.
`auth` contains values specific for the auth pods. You can use it to
set specific values for auth pods, taking precedence over chart-scoped values.

For example, to override the [`postStart`](#postStart) value only for auth pods:
```yaml
# By default all pods postStart command should be "echo starting"
postStart: ["echo", "starting"]

auth:
  # But we override the `postStart` value specifically for auth pods
  postStart: ["curl", "http://hook"]
  imagePullPolicy: Always
```

### `auth.teleportConfig`

| Type     | Default value | Required? |
|----------|---------------|-----------|
| `object` | {}            | no        |

`auth.teleportConfig` contains YAML teleport configuration for auth pods.

> The configuration will be merged with the chart-generated configuration

Reviewer 3: Is it worth it to clarify what the merge rules are? For example, if the field is a list, do these get merged or override?

Author: Yes it is, because I'm not 100% sure of the answer and it is not documented in sprig. I would expect the list to be overridden. I'll do some tests

| `standalone` | Runs by relying only on Kubernetes resources. | [Getting Started - Kubernetes with SSO](../../deploy-a-cluster/helm-deployments/kubernetes-cluster.mdx) |
| `aws` | Leverages AWS managed services to store data. | [Running an HA Teleport cluster using an AWS EKS Cluster](../../deploy-a-cluster/helm-deployments/aws.mdx) |
| `gcp` | Leverages GCP managed services to store data. | [Running an HA Teleport cluster using a Google Cloud GKE cluster](../../deploy-a-cluster/helm-deployments/gcp.mdx) |
| `scratch` (v12 and above) | Generates empty Teleport configuration. User must pass their own config. | [Running a Teleport cluster with a custom config](../../deploy-a-cluster/helm-deployments/custom.mdx) |

<Admonition type="note" title="Custom mode removal">

> `custom` mode has been removed in Teleport version 12. See the [version 12 migration guide]().

Reviewer 2: Add link!

</Admonition>

<Admonition type="warning">

> The `clusterName` cannot be changed during a Teleport cluster's lifespan.hugo/chart-split-docs

Reviewer 2: ```suggestion
The `clusterName` cannot be changed during a Teleport cluster's lifespan.
```

Random clipboard fart

The configuration will be merged with the chart-generated configuration
and will take precedence in case of conflict.

> The merge logic is the following:

Reviewer 2: ```suggestion
The merge logic is as follows:
```

The configuration will be merged with the chart-generated configuration
and will take precedence in case of conflict.

> The merge logic is the following:

Reviewer 2: ```suggestion
The merge logic is as follows:
```

### `authentication.secondFactor`

| Type | Default value | Required? | `teleport.yaml` equivalent |
| - | - | - | - |

> | `string` | `otp` | Yes | `auth_service.authentication.second_factor` |

Author: done

> | `string` | `otp` | Yes | `auth_service.authentication.second_factor` |

Reviewer: Thanks! I hope you didn't do this manually; VSCode has an extension for that.

## `aws`

`aws` settings are described in the AWS guide: [Running an HA Teleport cluster using an AWS EKS Cluster](../../deploy-a-cluster/helm-deployments/aws.mdx)

## `gcp`

`gcp` settings are described in the GCP guide: [Running an HA Teleport cluster using a Google Cloud GKE cluster](../../deploy-a-cluster/helm-deployments/gcp.mdx)

## `highAvailability`

`highAvailability` contains settings controlling how Teleport pods are

> replicated and scheduled. This allows to run Teleport in a highly-available

Reviewer: Sentence is missing a subject. "This allows ___ to run..."

Reviewer: Or alternately, "This allows Teleport to run..."

fashion: Teleport should sustain the crash/loss of a machine without interrupting
the service.

> **For auth pods**:

Reviewer: Suggest using headers instead of using bold as a header:

```suggestion
### For auth pods
```

or 

```suggestion
#### For auth pods
```

Author: Unresolving this because the linter is not happy.

Using heading level 3 would make the section appear in the lateral toc which only contains value names, so this would be confusing.

Using header level 4 is OK, but the linter does not like that the heading level 3 is skipped.

Do you know a trick to make the linter happy with a level 4 header, or to hide a level 3 header from the toc?

When using "standalone" or "scratch" mode, you must use highly-available storage
(etcd, DynamoDB or Firestore) for multiple replicas to be supported.
Manually configuring NFS-based storage or ReadWriteMany volume claims
is NOT supported and will result in errors. Using Teleport's built-in

> ACME client (opposed to using cert-manager or passing certs through a secret)

Reviewer: ```suggestion
ACME client (as opposed to using cert-manager or passing certs through a secret)
```

is not supported with multiple replicas.

> **For proxy pods**:

Reviewer: Same as above:

```suggestion
### For proxy pods
```

or:

```suggestion
#### For proxy pods
```

> Proxy pods need to be provided a certificate to be replicated (either via
> tls.existingSecretName` or via `highAvailability.certManager`).

Reviewer: Fixes inline code markers, and removes redundant `via`.

```suggestion
Proxy pods need to be provided a certificate to be replicated (via either
`tls.existingSecretName` or `highAvailability.certManager`).
```
